services/director-v2/src/simcore_service_director_v2/utils/dags.py

```diff
@@ -30,11 +39,64 @@ def create_dag_graph(workbench: Workbench) -> nx.DiGraph:
                     and to_node_class(predecessor_node.key) == NodeClass.COMPUTATIONAL
                 ):
                     dag_graph.add_edge(str(input_node_id), node_id)
-    log.debug("created DAG graph: %s", nx.to_dict_of_lists(dag_graph))
 
     return dag_graph
 
 
+def mark_node_dirty(graph: nx.DiGraph, node_id: NodeID):
+    graph.nodes()[str(node_id)]["dirty"] = True
+
+
+def is_node_dirty(graph: nx.DiGraph, node_id: NodeID) -> bool:
+    return graph.nodes()[str(node_id)].get("dirty", False)
+
+
+def _node_outdated(full_dag_graph: nx.DiGraph, node_id: NodeID) -> bool:
+    node = full_dag_graph.nodes(data=True)[str(node_id)]
+    # if the node has no output it is outdated for sure
+    if not node["outputs"]:
+        return True
+    for output_port in node["outputs"]:
+        if output_port is None:
+            return True
+    # ok so we have outputs, but maybe the inputs are old? let's check recursively
+    for input_port in node["inputs"]:
+        if isinstance(input_port, PortLink):
+            if is_node_dirty(full_dag_graph, input_port.node_uuid):
+                return True
+        else:
+            # FIXME: here we should check if the current inputs are the ones used to generate the current outputs
+            # this could be done by saving the inputs as metadata together with the outputs (see blockchain)
+            # we should compare the current inputs with the inputs used for generating the current outputs!!!
```

Comment:
Nice, you'd like to add a check where if the inputs did not change, there is no need to compute the results again?

Reply:
right.